src/__experimental__/components/decimal/decimal.component.js

```diff
@@ -0,0 +1,107 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import TextBox from '../textbox';
+import I18nHelper from '../../../utils/helpers/i18n';
+
+const Decimal = (props) => {
+  const [decimalValue, setDecimalValue] = useState('0.00');
+
+  useEffect(() => {
+    // Update decimalValue if precision prop changes
+    const { precision } = props;
+    const parts = decimalValue.split('.');
+    let newDecimal = parts[1];
+    const decimalLength = parts[1].length;
+    let diff;
+
+    if (precision > decimalLength) { // Precision increased
+      diff = precision - decimalLength;
+      for (let i = 0; i < diff; i++) {
+        newDecimal = newDecimal + '0';
+      }
+    } else if (precision < decimalLength) { // Precision decreased
+      diff = decimalLength - precision;
+      newDecimal = Math.round(newDecimal / (10 * diff));
+    }
+    
+    setDecimalValue([parts[0], '.', newDecimal].join(''));
+  }, [props.precision]);
+
+  const testValue = (value) => {
+    const { precision } = props;
+    const format = I18nHelper.format();
+    const delimiter = `\\${format.delimiter}`;
+    const seperator = `\\${format.separator}`;
+
+    const regex = new RegExp(`^[\\d${delimiter}]*[${seperator}{1}]?\\d{0,${precision}}?$`);
+    const isGoodDecimal = regex.test(value);
+
+    return isGoodDecimal;
+  }
+
+  const handleChange = (evt) => {
+    const target = evt.target;
+    const { value, selectionEnd } = evt.target;
+    const testString = testValue(value);
+    
+    if (testString) {
+      setDecimalValue(value);
+      props.onChange({ target: { value: value } })
+      setTimeout(() => {
+        target.setSelectionRange(selectionEnd, selectionEnd)
+      });
+    }
+  }
+
+  const handleBlur = () => {
+    const addCommasRegex = /(\d)(?=(\d{3})+(?!\d))/g;
+    const parts = decimalValue.split('.');
+    const integer = parts[0];
+    
+    let newInteger = integer.replace(/,/g, '');
+    newInteger = newInteger.replace(addCommasRegex, '$1,');
+    let array = [newInteger];
+
+    if (parts[1]) {
+      array.push('.', parts[1]);
+    }
+    setDecimalValue(array.join(''));
+  }
+
+  return (
+    <TextBox
+      {...props}
+      onChange={ handleChange }
+      onBlur={ handleBlur }
+      value={ decimalValue }
```

Comment:
this looks like the value is driven by state rather than by props (eg. if i change props.value will it update?) I know the original used state too, but it listened for changes to prop.value and updated accordingly. Maybe there is a way to do it without state at all? eg. derived data

Reply:
I've ended up sticking with storing the value in `state`. I couldn't figure out a way to do this with derived data, as there were too many instances where I needed to change the value in different ways (`value` in `onChange`, delimiters in `onBlur` and the length of the decimal when `precision` changes).

Using `state` also means that I always have the previous value easily available if a new value fails validation.